File: connect4/Connect4.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>

// template of empty spaces that shall be printed
// char ptemplate[6][8] = {
//                      {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '},
//                      {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '},
//                      {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '},
//                      {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '},
//                      {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '},
//                      {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '}
//                     };

int header[7] = {1, 2, 3, 4, 5, 6, 7};

typedef char player;
player p1 = 'x';
player p2 = 'o';

int one = 0, two = 0, three = 0, four = 0, five = 0, six = 0, seven = 0;

char** createBoard();
void emptyBoard(char** ptemplate);
void initialize(char** ptemplate);
bool scanTemplate(char** ptemplate, int playerNum, player p);
bool scanBoard(char** ptemplate);
bool scanRow(char** ptemplate, player p);
bool scanColumn(char** ptemplate, player p);
bool scanDiag(char** ptemplate, player p);
void choosePlayer(char** ptemplate, player p);
void playMove(char** ptemplate, player p);
void go(char** ptemplate, int m, int col, player p, int pNum);

int main() {
    char** ptemplate = createBoard();
    initialize(ptemplate); // prints the empty board
    playMove(ptemplate, p1);
}

char** createBoard() {
  char** board = malloc(sizeof(char*) * 6);
  for (int i = 0; i < 6; i++) {
    board[i] = malloc(sizeof(char*) * 7);
  }
  for (int i = 0; i < 6; i++) {
    for (int j = 0; j < 8; j++) {
      board[i][j] = ' ';
    }
  } return board;
}

void emptyBoard(char** ptemplate) {
  for (int i = 0; i < 6; i++) {
        for (int j = 0; j < 8; j++) {
            ptemplate[i][j] = ' ';
        }
  }
  one = 0; two = 0; three = 0; four = 0; five = 0; six = 0; seven = 0;
}

void initialize(char** ptemplate) {
    for (int i = 0; i < 7; i++) {
        printf(" %d", header[i]);
    } printf("\n");
    for (int j = 0; j < 6; j++) {
        for (int i = 0; i < 8; i++) {
            printf("|%c", ptemplate[j][i]);
        } printf("\n");
    }
    printf("```````````````");
    printf("\n");
}

void playAgain(char** ptemplate) {
    char yesNo;
    scanf("%c", &yesNo);
    // fgets(&yesNo, 1, stdin);
    yesNo = getchar();
    if (yesNo == 'y') {
      system("clear");
      emptyBoard(ptemplate);
      initialize(ptemplate);
      playMove(ptemplate, p1);
    } else if (yesNo == 'n') {
      free(ptemplate);
      exit(0);
    } else {
        printf("Sorry, response not valid. Please enter again.\n");
        playAgain(ptemplate);
    }
}

bool scanTemplate(char** ptemplate, int playerNum, player p) {
    bool notFull = scanBoard(ptemplate);
    bool fourInARow = scanRow(ptemplate, p);
    bool fourInACol = scanColumn(ptemplate, p);
    bool fourInADiag = scanDiag(ptemplate, p);
    if (notFull == true && fourInARow == false && fourInACol == false && fourInADiag == false) {
        initialize(ptemplate);
        return true;
    } else if (notFull == false) {
        initialize(ptemplate);
        printf("Board is full, play again? (y/n)\n");
        playAgain(ptemplate);
        return false;
    } else if (fourInARow == true || fourInACol == true || fourInADiag == true) {
        initialize(ptemplate);
        printf("Player %d has won. Play again? (y/n)\n", playerNum);
        playAgain(ptemplate);
        return false;
    }
}

// this scans the board to see if there are any empty spaces to be filled
// if there are no empty spaces (isEmpty == 0) game is terminated
bool scanBoard(char** ptemplate) {
    int numEmpty = 0;
    for (int i = 0; i < 6; i++) {
        for (int j = 0; j < 7; j++) {
            if (ptemplate[i][j] == ' ') {
                numEmpty++;
            }
        }
    }
    if (numEmpty > 0) {
        return true;
    } else if (numEmpty == 0) {
        return false;
    }
}

bool scanRow(char** ptemplate, player p){
  for (int y = 0; y < 7; y++) {
    for (int x = 0; x < 3; x++) {
      if (ptemplate[x][y] == p && ptemplate[x+1][y] == p && ptemplate[x+2][y] == p && ptemplate[x+3][y] == p) {
        return true;
      }
    }
  }
  return false;
}

bool scanColumn(char** ptemplate, player p) {
  for (int y = 0; y < 4; y++) {
    for (int x = 0; x < 6; x++) {
      if (ptemplate[x][y] == p && ptemplate[x][y+1] == p && ptemplate[x][y+2] == p && ptemplate[x][y+3] == p) {
        return true;
      }
    }
  }
  return false;
}

bool scanDiag(char** ptemplate, player p) {
  for (int y = 3; y < 7; y++) {
    for (int x = 0; x < 3; x++) {
      if (ptemplate[x][y] == p && ptemplate[x+1][y-1] == p && ptemplate[x+2][y-2] == p && ptemplate[x+3][y-3] == p) {
        return true;
      }
    }
  }

  for (int y = 0; y < 4; y++) {
    for (int x = 0; x < 3; x++) {
      if (ptemplate[x][y] == p && ptemplate[x+1][y+1] == p && ptemplate[x+2][y+2] == p && ptemplate[x+3][y+3] == p) {
        return true;
      }
    }
  }
  return false;
}

void choosePlayer(char** ptemplate, player p) {
    if (p == p1) {
        p = p2; // toggles
        playMove(ptemplate, p);
    } else if (p == p2) {
        p = p1; // toggles
        playMove(ptemplate, p);
    }
}

void playMove(char** ptemplate, player p) {
    int theInput;
    int playerNum;

    switch(p) {
        case 'x':
            playerNum = 1; // translates so that it's easier to print out
            break;
        case 'o':
            playerNum = 2;
            break;
    }

    printf("Player %d, choose column to play.\n", playerNum);
    if(scanf("%d", &theInput)) {
        switch(theInput) {
          case 1:
            {
              one++;
              go(ptemplate, one, 1, p, playerNum); // move is executed
              break;
            }
          case 2:
            {
              two++;
              go(ptemplate, two, 2, p, playerNum); // move is executed
              break;
            }
          case 3:
            {
              three++;
              go(ptemplate, three, 3, p, playerNum); // move is executed
              break;
            }
          case 4:
            {
              four++;
              go(ptemplate, four, 4, p, playerNum); // move is executed
              break;
            }
          case 5:
            {
              five++;
              go(ptemplate, five, 5, p, playerNum); // move is executed
              break;
            }
          case 6:
            {
              six++;
              go(ptemplate, six, 6, p, playerNum); // move is executed
              break;
            }
          case 7:
            {
              seven++;
              go(ptemplate, seven, 7, p, playerNum); // move is executed
              break;
            }
          default:
              printf("ERROR: invalid column number.\n");
              system("clear");
              initialize(ptemplate);
              playMove(ptemplate, p); // player not penalized for entering an invalid column number
          }
    } else {
      printf("Error: input not a number\n");
      initialize(ptemplate);
      playMove(ptemplate, p);
    }
}

void go(char** ptemplate, int m, int col, player p, int pNum) {
  if (m > 5){ 
    printf("Column full.");
    system("clear");
    bool cont = scanTemplate(ptemplate, pNum, p);
    if (cont == true){
        playMove(ptemplate, p);
    } else if (cont == false) {
        free(ptemplate);
        exit(0);
    }
  } else {
    ptemplate[6 - m][col-1] = p;
    system("clear");
    bool cont = scanTemplate(ptemplate, pNum, p);
    if (cont == true){
        choosePlayer(ptemplate, p); // next player
    } else if (cont == false) {
        free(ptemplate);
        exit(0);
    } // new board will be printed
  }
}

File: connect4/Connect4.py
from time import sleep
import sys
import os

def make_board():
  board = [[' ', ' ', ' ', ' ', ' ', ' '], 
          [' ', ' ', ' ', ' ', ' ', ' '], 
          [' ', ' ', ' ', ' ', ' ', ' '], 
          [' ', ' ', ' ', ' ', ' ', ' '], 
          [' ', ' ', ' ', ' ', ' ', ' '], 
          [' ', ' ', ' ', ' ', ' ', ' '], 
          [' ', ' ', ' ', ' ', ' ', ' ']]
  return board

p1 = 'x'
p2 = 'o'

board = make_board()

one = [];
two = [];
three = [];
four = [];
five = [];
six = [];
seven = [];

def empty():
  for row in range(len(board[0])):
    for col in range(len(board)):
      board[col][row] = " "

def initialize():
  header = ' '
  for num in range(1, len(board) + 1):
    header += ' '+ str(num) +'  '
  print(header)

  print('+---' * (len(board)) + '+')

  # Printing each row of the board
  for row in range(len(board[0])):
    row_with_items = ''
    for col in range(len(board)):
      row_with_items += ('| '+str(board[col][row])) + ' '
    print(row_with_items + '|')
    print('+---' * (len(board)) + '+')

def scanTemplate(playerNum, p):
  notFull = scanBoard()
  fourInARow = scanRow(p)
  fourInACol = scanColumn(p)
  fourInADiag = scanDiag(p)
  if (notFull == True and fourInARow == False and fourInACol == False and fourInADiag == False):
    initialize()
    return True
  elif (fourInARow == True or fourInACol == True or fourInADiag or True):
    initialize()
    print("Player " + playerNum + " has won.")
    sleep(1)
    playAgain()  
    return False
  elif notFull == False:
    initialize()
    print("Board is full")
    sleep(1);
    playAgain()  
    return False

def playAgain():
  print("Play again?")
  yn = input()
  yesno = yn.lower()
  if (yesno == "yes"):
    clear = lambda: os.system('clear')
    clear()
    empty()
    initialize()
    playMove(p1)
  else:
    sys.exit("~~yeet~~")

def scanBoard():
  numEmpty = 0;
  for j in range(5):
    for i in range(6):
      if board[i-1][j-1] == " ":
        numEmpty += 1
  if numEmpty > 0:
    return True
  elif numEmpty == 0:
    return False

def scanRow(p):
  for y in range(len(board[0])):
    for x in range(len(board) - 3):
      if (board[x][y] == p and board[x+1][y] == p and board[x+2][y] == p and board[x+3][y] == p):
        return True
  return False

def scanColumn(p):
  for x in range(len(board)):
    for y in range(len(board[0]) - 3):
      if (board[x][y] == p and board[x][y+1] == p and board[x][y+2] == p and board[x][y+3] == p):
        return True
  return False

def scanDiag(p):
  for x in range(len(board) - 3):
    for y in range(3, len(board[0])):
      if board[x][y] == p and board[x+1][y-1] == p and board[x+2][y-2] == p and board[x+3][y-3] == p:
        return True
  return False

  for x in range(len(board) - 3):
    for y in range(len(board[0]) - 3):
      if board[x][y] == p and board[x+1][y+1] == p and board[x+2][y+2] == p and board[x+3][y+3] == p:
        return True
  return False

def choosePlayer(p):
  if p == p1:
    p = p2
    playMove(p)
    return p
  elif (p == p2):
    p = p1;
    playMove(p)
    return p

def playMove(p):
  choice = -1
  theInput = ""
  playerNum = 0
  if (p == 'x'):
    playerNum = "one"
  elif (p == 'o'):
    playerNum = "two"
  print("Player " + playerNum + ", choose column to play.")
  theInput = input()

  try:
    val = int(theInput)
  except ValueError:
    print("not an int")
    sleep(0.50)
    clear = lambda: os.system('clear')
    clear()
    # print ("\n" * 100)
    initialize();
    playMove(p);

  choice = int(theInput)

  if (choice == 1):
    one.append(0)
    maxNum = max(one)
    one.append(maxNum + 1)
    go(maxNum, 1, p, playerNum)
  elif (choice == 2):
    two.append(0)
    maxNum = max(two)
    two.append(maxNum + 1)
    go(maxNum, 2, p, playerNum)
  elif (choice == 3):
    three.append(0)
    maxNum = max(three)
    three.append(maxNum + 1)
    go(maxNum, 3, p, playerNum)
  elif (choice == 4):
    four.append(0)
    maxNum = max(four)
    four.append(maxNum + 1)
    go(maxNum, 4, p, playerNum)
  elif (choice == 5):
    five.append(0)
    maxNum = max(five)
    five.append(maxNum + 1)
    go(maxNum, 5, p, playerNum)
  elif (choice == 6):
    six.append(0)
    maxNum = max(six)
    six.append(maxNum + 1)
    go(maxNum, 6, p, playerNum)
  elif (choice == 7):
    seven.append(0)
    maxNum = max(seven)
    seven.append(maxNum + 1)
    go(maxNum, 7, p, playerNum)
  else:
    print("out of bounds")
    sleep(0.50)
    clear = lambda: os.system('clear')
    clear()
    # print ("\n" * 100)
    initialize();
    playMove(p);

def go(m, col, p, pNum):
  if (m > 5):
    print("Column full.")
    sleep(0.50)
    clear = lambda: os.system('clear')
    clear()
    # print ("\n" * 100)
    cont = scanTemplate(pNum, p)
    if (cont == True):
      playMove(p)
    elif (cont == False):
      playAgain()  
  else:
    board[col - 1][5 - m] = p;
    clear = lambda: os.system('clear')
    clear()
    # print ("\n" * 100)
    cont = scanTemplate(pNum, p)
    while (cont == True):
      p = choosePlayer(p)
      playMove(p)
    playAgain()
    sys.exit("~~yeet~~")  

initialize()
playMove(p1)
